keps/draft-kuttl-as-a-library.md

```diff
@@ -0,0 +1,193 @@
+* Allow for "breaking the glass" when declarative tests are not expressive enough.
+
+### Non-Goals
+
+* Creating a new spin-off testing project.
+* Fragmentation of our operator developer community.
+* Dropping support for YAML-based declarative tests.
+
+### Principles
+
+* Useful logging by default
+* Depend on existing libraries whenever possible
+* Retrying by default where it makes sense
+* Ease of adding retrying to non-retried functions
+* It is better to be consistent than to be "perfect"
+* Err on the side of flexibility
+* Err on the side of returning errors instead of panicking/Fatalf/etc.
+* Cohesive logging output (parameters, formatting, etc.)
+* Separation of concerns between domains (KUDO, Kubernetes, etc.)
+
+## Proposal
+
+We will improve and break-down the interface of the kuttl library to make it easier to consume as a downstream developer. Libraries that need to be broken out of kuttl:
+
+* A generic Kubernetes client (currently this is contained inside of `pkg/test/harness.go`).
+* Kubernetes object constructors (this is already pretty good in `pkg/test/utils/kubernetes.go`).
+* Assertion methods (needs better documentation for library usage).
+* Test setup and tear down methods (is currently coupled to steps).
+* Cluster and test environment provisioning (this is currently coupled to `Harness` in `harness.go`).
+
+Additionally, KUDO-specific helper methods will be provided in the library that can be used by library users.
+
+We will also endeavor to stay close to the API used by the kudo-cassandra-operator testing library to ensure that migration of existing tests is easy.
+
+### cmd/cmd.go
+
+The cmd package should provide conveniences for running arbitrary commands (i.e. shelling out) from Go code in integration tests.
+
+Current functionality:
+* Providing an Exec function that returns exit status, stdout and stderr
+
+```
+package cmd
+
+func Exec(
+  command string,
+  arguments []string,
+  environmentVariables map[string]string,
+) (
+  exitStatus int,
+  stdout *bytes.Buffer,
+  stderr *bytes.Buffer,
+  error error,
+)
+```
+
+### kubectl/kubectl.go
+
+The kubectl package should provide conveniences for interacting with a Kubernetes cluster as one would do via kubectl from the shell.
+
+```
+package kubectl
+
+func Exec(
+  namespaceName string,
+  containerName string,
+  command string,
+  argvuments []string,
+  environmentVariables map[string]string,
+) error
+```
+
+### k8s/k8s.go
+
+The k8s package should provide conveniences for interacting with Kubernetes objects.
+
+* Partial CRUD functionality for namespaces
+* CRUD functionality for all useful Kubernetes objects
+* Current known use-cases:
+* Creating and deleting namespaces during test runs
+
+```
+package k8s
+
+func Init(kubectlOptions *kubectl.KubectlOptions) error
+
+func CreateNamespace(namespaceName string) error
+
+func DeleteNamespace(namespaceName string) error
+```
+
+### kudo/kudo.go
```

Comment:
The question is, how closely related to KUDO do we want to keep KUTTL? I think we have the plan to remove the KUDO-specific parts from the test harness, and allow kudo to plug into it instead, do we want the same for the library parts?

Reply:
There currently is a need for KUDO-specific part because KUDO doesn't provide a consistent API to use it as a library (somewhat captured in https://github.com/kudobuilder/kudo/issues/1002). While it has a versioned clientset, other functionality, like installing packages is only provided in packages without any compatibility guarantees and should currently be considered internal.

Reply:
Yeah, while we're decoupling the CLI itself (in order to support more projects), I have no problem with providing helper methods in a special package for kuttl - helpers are cheap. We could also put them in the kudo repo (or even maybe a kudo-client repo?). Not too particular where that part of this goes.

Reply:
BTW, KEP for decoupling is here: https://github.com/kudobuilder/kuttl/pull/3

Reply:
Agreed. I think the ideal is for `kuttl` to be completely standalone and in-and-of-itself not have any knowledge of KUDO. That's when we succeed with this KEP. For today, if there's cross references, that's OK, but part of this KEP IMO should include a plan to create APIs that specifically detangle the dependency, or at least trigger KUDO to have a kuttl package of it's own for additional helpers.